standardization/src/main/scala/za/co/absa/enceladus/standardization/interpreter/stages/newVer/TypeParser.scala

```diff
@@ -0,0 +1,370 @@
+import za.co.absa.enceladus.standardization.interpreter.dataTypes.ParseOutput
+import za.co.absa.enceladus.utils.schema.SchemaUtils
+import za.co.absa.enceladus.utils.schema.SchemaUtils.appendPath
+import za.co.absa.enceladus.utils.types.Defaults
+import org.apache.spark.sql.functions._
+import za.co.absa.spark.hofs.transform
+import za.co.absa.enceladus.utils.error.ErrorMessage
+import za.co.absa.enceladus.utils.time.DateTimePattern
+
+import scala.util.Random
+
+/**
+  * Base trait for standardization function
+  * Each final class in the hierarchy represents a `standardize` function for its specific data type field
+  * Class hierarchy:
+  *   TypeParser
+  *     ArrayParser !
+  *     StructParser !
+  *     PrimitiveParser
+  *       SimpleParser
+  *         NumericParser !
+  *         StringParser !
+  *         BooleanParser !
+  *       DateTimeParser
+  *         TimestampParser !
+  *         DateParser !
+  */
+sealed trait TypeParser {
+  def standardize(): ParseOutput
+
+  val origSchema: StructType
+  val field: StructField
+  val path: String
+  val parent: TypeParser.Parent
+  val fieldName: String = SchemaUtils.getFieldNameOverriddenByMetadata(field)
+  val currentAttrPath: String = appendPath(path, fieldName) // absolute path to field to parse
+  val column: Column = parent.map {
+    case Left(x) => x //parent is array
+    case Right(x) => x(fieldName) //parent is struct
+  }.getOrElse(col(currentAttrPath)) // no parent
+  val isArrayElement: Boolean = parent.exists(_.isLeft)
+}
+
+object TypeParser {
+  private val logger: Logger = LoggerFactory.getLogger(this.getClass)
+
+  def standardize(field: StructField, path: String, origSchema: StructType): ParseOutput = {
+    TypeParser(field, path, origSchema).standardize()
+  }
+
+  private type Parent = Option[Either[Column, Column]]
+
+  private object Parent {
+    def array(column: Column): Parent = Some(Left(column))
+    def struct(column: Column): Parent = Some(Right(column))
+  }
+
+  private def apply(field: StructField, path: String, origSchema: StructType, parent: Parent = None): TypeParser = {
+    val parserClass = field.dataType match {
+      case _: ArrayType     => ArrayParser
+      case _: StructType    => StructParser
+      case _: NumericType   => NumericParser
+      case _: StringType    => StringParser
+      case _: BooleanType   => BooleanParser
+      case _: DateType      => DateParser
+      case _: TimestampType => TimestampParser
+      case t => throw new IllegalStateException(s"${t.typeName} is not a supported type in this version of Enceladus")
+    }
+    parserClass(field, path, origSchema, parent)
+  }
+
+  private final case class ArrayParser(field: StructField,
+                                       path: String,
+                                       origSchema: StructType,
+                                       parent: Parent) extends TypeParser {
+    private val fieldType: ArrayType = field.dataType.asInstanceOf[ArrayType]
```

Comment:
This could probably be left to type inference, since we explicitly cast it to `ArrayType`